### ISSUE TITLE
Adding stan math c++ to conda forge ?

### ISSUE THREAD

**Issue by faysou** [stan-dev/math#2581]
This is more a feature request than a bug.

Would it be possible to add stand math c++ to conda-forge ? 
https://conda-forge.org/

This way it would be possible to work through an easy installation with xeus-cling, which is a c++ interpreter in a jupyter notebook.
https://github.com/jupyter-xeus/xeus-cling

Thank you !

**Comment by bob-carpenter** [comment]
Thanks. This is the right place for feature requests, too.  I don't imagine it'll be hard because stan-dev/math gets installed as part of CmdStan.  I'm pinging @WardBrian, who worked on putting CmdStan on conda.

**Comment by WardBrian** [comment]
This is definitely reasonable (and probably the 'right way' to handle CmdStan on conda is to do this instead, assuming we can point the CmdStan makefiles at the separate math lib), I just wasn't sure if people would it separately from CmdStan

**Comment by WardBrian** [comment]
Just wondering @faysou - what would you be looking for out of a conda installation? stan-math doesn't have something like a `make install` target. It seems like mainly what it would offer is an easy way to download the package and put it somewhere on the library path, but not much else. Am I missing some benefit?

**Comment by bob-carpenter** [comment]
I don't know how they're built, but I think there are at least three binaries required:

* Intel Thread Building Blocks (low-level threading tools), 
* Sundials (ode solver), and
*  MPI (multi-process).

**Comment by faysou** [comment]
@WardBrian what I would like is to work interactively with stan-math inside a jupyter c++ notebook. This is possible with xeus-cling. I'm curious if automatic differentiation would also work in the notebook interactively.

**Comment by WardBrian** [comment]
I suppose my question is, what about the current process is not able to support that which a conda installation would? Are you currently able to point xeus-cling at a local copy of stan-math?

**Comment by faysou** [comment]
No, I haven't even tried. The point of having stan-math as a conda-forge recipe would be to have conda install it without a local installation. Some projects use conda-forge for this purpose, so I'm point this out as I think it could be useful.

**Comment by bob-carpenter** [comment]
The question is whether cling would be up to compiling Stan and all its associated packages, some of which are binaries that need to be linked or somehow also compiled with cling.

**Comment by faysou** [comment]
I remember that the conda-forge project is designed to allow remote compilations for different target systems, so then users only download compiled binaries for their system.

**Comment by WardBrian** [comment]
That is correct, but there is no compiled binary for stan-math

Several of the libraries it depends on are already on conda, such as tbb here: https://anaconda.org/conda-forge/tbb

I'm pretty certain only macos builds use clang however, so I'm not sure how compatible with xeus-cling they end up being

**Comment by faysou** [comment]
I'm not sure of the feasibility, I don't know stan-math enough. My suggestion was more from a user point of view, it could make prototyping easier.

**Comment by WardBrian** [comment]
I'm not terribly familiar with C++ library packaging, but my understanding [header-only](https://en.wikipedia.org/wiki/Header-only) libraries like stan-math have no object file they get compiled to. This is different from packages like boost (just picking an example off your list above), some parts of which get installed by building a `.(dylib|so|dll)` file and then placing that somewhere.

To phrase this another way, if you've downloaded the stan-math repository, I believe that you  _have_ installed it. The only difficulty here is that it depends on other libraries which do need to be built, but those are likely available on conda.

I think all you would need to do is download stan-math, install tbb and sundials from conda, point stan-math to those libraries in `make/local`, and then point cling to where stan-math is downloaded and the #includes should work. I don't think conda can make this any simpler besides perhaps pre-setting the `make/local` variables unless there is something I don't understand here. 

I was able to find a few header-only libraries on conda, but they are only there if they use cmake to set up dependency paths so other people can link to them. 

@bob-carpenter probably knows better if there really is a sense in which stan-math can be "installed" other than just downloading it.

**Comment by faysou** [comment]
@WardBrian, then what you say means a conda recipe on conda forge is possible.

**Comment by WardBrian** [comment]
I don’t think there’s any meaningful way of having stan-math on conda the way it is currently conceived of. If we added a cmake installation method to the library then we could make a very minimal conda package which basically consisted of the output of that process

**Comment by faysou** [comment]
I'll try your suggestion then. You are better placed than me to know what's possible with the library. Maybe it will be possible in the future, at least it's on your radar. Thank you

**Comment by WardBrian** [comment]
Again, @bob-carpenter knows more and may conceive of better way of installing. 

You could also try installing the cmdstan package on conda, as it contains stan-math and the required libraries, and then pointing cling towards <CONDA_ENV>/bin/cmdstan/stan/lib/stan_math

**Comment by faysou** [comment]
Ok thank you. 
Here's another similar discussion for packaging in conda a c++ library.

https://github.com/Dobiasd/FunctionalPlus/issues/161#issuecomment-576007525

**Comment by bob-carpenter** [comment]
The parts of Stan math that the Stan team wrote are largely header only.  The dependent libraries are not.  I'm not sure how TBB, Sundials, and MPI get built as part of the Stan math lib.  There's nothing in Stan per se to compile as far as I know, just these external dependencies.  

Somehow, these libraries must be getting either built or just brought in as binaries with the CmdStan install.

**Comment by WardBrian** [comment]
FunctionalPlus uses cmake to generate an install target for conda to use, like the other header libraries I mentioned

The libraries in stan that need to be built are build during the cmdstan build, or you can point to existing versions with a make variable. In the cmdstan conda build, we do a bit of both. I’m not sure if stan-math by itself even has a make target to build them or if it’s all upstream
